File: test/arguments/with_range_check_code/python/VariableBitRangeCheckTest.py
import unittest
import zserio

from testutils import getZserioApi

class VariableBitRangeCheckTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "with_range_check_code.zs",
                               extraArgs=["-withRangeCheckCode"]).variable_bit_range_check

    def testVariableBitLowerBound(self):
        self._checkVariableBitValue(VARIABLE_BIT_LOWER_BOUND)

    def testVariableBitUpperBound(self):
        self._checkVariableBitValue(VARIABLE_BIT_UPPER_BOUND)

    def testVariableBitBelowLowerBound(self):
        with self.assertRaises(zserio.PythonRuntimeException):
            self._checkVariableBitValue(VARIABLE_BIT_LOWER_BOUND - 1)

    def testVariableBitAboveUpperBound(self):
        with self.assertRaises(zserio.PythonRuntimeException):
            self._checkVariableBitValue(VARIABLE_BIT_UPPER_BOUND + 1)

    def _checkVariableBitValue(self, value):
        variableBitRangeCheckCompound = self.api.VariableBitRangeCheckCompound.fromFields(NUM_BITS, value)
        writer = zserio.BitStreamWriter()
        variableBitRangeCheckCompound.write(writer)
        reader = zserio.BitStreamReader(writer.getByteArray())
        readVariableBitRangeCheckCompound = self.api.VariableBitRangeCheckCompound.fromReader(reader)
        self.assertEqual(variableBitRangeCheckCompound, readVariableBitRangeCheckCompound)

NUM_BITS = 10
VARIABLE_BIT_LOWER_BOUND = zserio.bitfield.getBitFieldLowerBound(NUM_BITS)
VARIABLE_BIT_UPPER_BOUND = zserio.bitfield.getBitFieldUpperBound(NUM_BITS)

File: test/arguments/with_range_check_code/cpp/VariableBitRangeCheckTest.cpp
#include "gtest/gtest.h"

#include "zserio/CppRuntimeException.h"

#include "with_range_check_code/variable_bit_range_check/VariableBitRangeCheckCompound.h"

namespace with_range_check_code
{
namespace variable_bit_range_check
{

class VariableBitRangeCheckTest : public ::testing::Test
{
protected:
    void checkVariableBitValue(uint64_t value)
    {
        VariableBitRangeCheckCompound variableBitRangeCheckCompound;
        variableBitRangeCheckCompound.setNumBits(NUM_BITS);
        variableBitRangeCheckCompound.setValue(value);
        zserio::BitStreamWriter writer;
        variableBitRangeCheckCompound.write(writer);
        size_t writeBufferByteSize;
        const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
        zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
        const VariableBitRangeCheckCompound readVariableBitRangeCheckCompound(reader);
        ASSERT_EQ(variableBitRangeCheckCompound, readVariableBitRangeCheckCompound);
    }

    static const uint8_t NUM_BITS;
    static const uint64_t VARIABLE_BIT_LOWER_BOUND;
    static const uint64_t VARIABLE_BIT_UPPER_BOUND;
};

const uint8_t VariableBitRangeCheckTest::NUM_BITS = 10;
const uint64_t VariableBitRangeCheckTest::VARIABLE_BIT_LOWER_BOUND = UINT64_C(0);
const uint64_t VariableBitRangeCheckTest::VARIABLE_BIT_UPPER_BOUND = (UINT64_C(1) << NUM_BITS) - 1;

TEST_F(VariableBitRangeCheckTest, variableBitLowerBound)
{
    checkVariableBitValue(VARIABLE_BIT_LOWER_BOUND);
}

TEST_F(VariableBitRangeCheckTest, variableBitUpperBound)
{
    checkVariableBitValue(VARIABLE_BIT_UPPER_BOUND);
}

TEST_F(VariableBitRangeCheckTest, variableBitAboveUpperBound)
{
    try
    {
        checkVariableBitValue(VARIABLE_BIT_UPPER_BOUND + 1);
        FAIL() << "Actual: no exception, Expected: zserio::CppRuntimeException";
    }
    catch (const zserio::CppRuntimeException& excpt)
    {
        ASSERT_STREQ("Value 1024 of VariableBitRangeCheckCompound.value exceeds the range of <0..1023>!",
                excpt.what());
    }
}

} // namespace variable_bit_range_check
} // namespace with_range_check_code
